### PR TITLE
Add offscreen dom parsing sample

### PR DESCRIPTION
**Pull request by sebastianbenz**

### CODE REVIEW

**Comment by sebastianbenz** [comment]
Hey @oliverdunk, I've played around with using offscreen docs for dom parsing. Would be great if you could take a look whether this is a good and idiomatic way to do this. There are a few things I'm not sure about, I'll flag these separately.

File: functional-samples/cookbook.offscreen-dom/background.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Close any existing offscreen documents to avoid inconsistencies
+// with an updated service worker version
+addEventListener('install', async (event) => {
+  closeOffscreenDocument();
```
Comment (sebastianbenz) [inline functional-samples/cookbook.offscreen-dom/background.js:18]: @oliverdunk not sure if this is really needed, but there were cases (I think) in which sw and offscreen doc were out of sync.

File: functional-samples/cookbook.offscreen-dom/background.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    data
+  });
+}
+
+async function closeOffscreenDocument() {
+  if (!(await chrome.offscreen.hasDocument())) {
+    return;
+  }
+  console.log('closing offscreen doc');
+  await chrome.offscreen.closeDocument();
```
Comment (sebastianbenz) [inline functional-samples/cookbook.offscreen-dom/background.js:75]: Not sure if this is a good pattern. As extensions can only have a single offscreen document they need to be shared across all extension features. Theoretically, two processing steps could run in parallel in the same offscreen doc.

Comment (sebastianbenz) [reply functional-samples/cookbook.offscreen-dom/background.js:75]: cc @oliverdunk

File: functional-samples/cookbook.offscreen-dom/background.js (hunk 1 above)
Comment (oliverdunk) [reply functional-samples/cookbook.offscreen-dom/background.js:18]: I haven't seen/done this before, but given offscreen documents are new it's possible there's an edge none of us have noticed. I'd be curious if you have steps to reproduce or a video of what you're seeing when this isn't present.

I haven't personally done any testing around extension updates so it's possible the offscreen documents doesn't get removed I guess.

File: functional-samples/cookbook.offscreen-dom/background.js (hunk 2 above)
Comment (oliverdunk) [reply functional-samples/cookbook.offscreen-dom/background.js:75]: I think closing after you've done work is a good pattern. Per https://github.com/GoogleChrome/developer.chrome.com/issues/5437, it's not a good idea to use `hasDocument` - there are a few different workarounds with different tradeoffs.

If an extension is likely to be running multiple things in offscreen documents at once, I honestly think the only way to be sure things are safe is to implement some sort of scheduler/queue that makes sure only one thing is being done at once.

**Review by oliverdunk — COMMENTED** [review]
Left some thoughts, overall this seems like what I'd expect though :)

**Comment by sebastianbenz** [comment]
@oliverdunk clients.matchAll did the trick - thanks for the suggestion!
@IanStanion-google I made a small tweak to your clipboard sample

PTAL.

File: functional-samples/cookbook.offscreen-dom/README.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
+This recipe shows how to use DOMParser in an Extension Service Worker using the [Offscreen API][1].
+
+## Context
+
+Extension Service Workers don't have direct access to the DOM. This example demonstrates how to use an
+offscreen document for parsing and modifying DOM from an Extension Service Worker. Offscreen document
```
Comment (oliverdunk) [inline functional-samples/cookbook.offscreen-dom/README.md:6]:
```suggestion
offscreen document for parsing and modifying the DOM from an Extension Service Worker. Offscreen documents
```

File: functional-samples/cookbook.offscreen-dom/README.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+and Extension Service Worker exchange data using message passing.
```
Comment (oliverdunk) [inline functional-samples/cookbook.offscreen-dom/README.md:7]:
```suggestion
and Extension Service Workers exchange data using message passing.
```

File: functional-samples/cookbook.offscreen-dom/README.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+## Running this extension
+
+1. Clone this repository.
+2. Load this directory in Chrome as an [unpacked extension][2].
+3. Open the Extension menu and click the extension named "Offscreen API - DOM Parsing".
+
+If you inspect the Extension Service Worker in Chrome DevTools, you can see thatk
```
Comment (oliverdunk) [inline functional-samples/cookbook.offscreen-dom/README.md:15]:
```suggestion
If you inspect the Extension Service Worker in Chrome DevTools, you can see that:
```

**Review by oliverdunk — CHANGES_REQUESTED** [review]
Left a couple of comments, but the code here looks great. I think you may have accidentally committed a popup and a bookmarks sample? Let me know when those are fixed and I'll approve :)

**Comment by sebastianbenz** [comment]
Thanks for the review! Fixed the readme and removed the accidentally committed files.

**Review by oliverdunk — APPROVED** [review]
LGTM 👍

**Pull request merged**